### DISCUSSION TITLE
SSE and Json

### DISCUSSION THREAD

**Discussion by karapy** [bigskysoftware/htmx#947] [Q&A]
Hi there,

I very new so please excuse me if my question may be wrong or strange
I am sending currency rates from backend (FASTAPI) by using SSE and in data I am sending json format data it's like:

`[{'symbol': 'BTCUSDT', 'price': '21418.71000000'}, {'symbol': 'DOGEUSDT', 'price': '0.07382000'}]`

in the test.html file I used below code 

```
 <div hx-sse="connect:/ticker swap:symbol" class="container">
   
    </div>
```

the result is realtime but i see whole json data"
`[{'symbol': 'BTCUSDT', 'price': '21418.71000000'}, {'symbol': 'DOGEUSDT', 'price': '0.07382000'}]`

how can I fix this and be able to whole currency rate or any kind of json data online?

**Comment by Renerick** [comment]
So, there are two ways to approach this.

The intended way to use SSE in htmx is to send HTML fragments instead of JSON (just like you would with normal requests). See the examples here https://htmx.org/extensions/server-sent-events/.

However, if you have to use JSON, there is a way around it by using client side templates extension https://htmx.org/extensions/client-side-templates/. Just add the attribute with templating engine of your choice to the element that connects to SSE endpoint and you should be good to go.

Also, hx-sse is not quite supported and kept for backward compatibility. You should use SSE extension https://htmx.org/extensions/server-sent-events/ instead (migration guide is at the bottom of the page)

**Reply by karapy** [reply]
> So, there are two ways to approach this.
> 
> The intended way to use SSE in htmx is to send HTML fragments instead of JSON (just like you would with normal requests). See the examples here https://htmx.org/extensions/server-sent-events/.
> 
> However, if you have to use JSON, there is a way around it by using client side templates extension https://htmx.org/extensions/client-side-templates/. Just add the attribute with templating engine of your choice to the element that connects to SSE endpoint and you should be good to go.
> 
> Also, hx-sse is not quite supported and kept for backward compatibility. You should use SSE extension https://htmx.org/extensions/server-sent-events/ instead (migration guide is at the bottom of the page)

I am not frontend designers but HTMX was just very interesting for me, I wonder to know for showing real time stock market price or crypto market price on a page what is best to use, SSE is a good choice?

**Reply by Renerick** [reply]
Yes SSE is a good choice for real time data

**Reply by karapy** [reply]
so as I understand it's no issue if I instead of just sending message via sse i send json data and it's completely not a wrong solution
then i can use that json data using  client-side-templates Extension i can show my data,
is it right?
please excuse me  i am newby and these are new for me

**Reply by Renerick** [reply]
Yes, this is perfectly good solution.

You would do it something like this 

```html
<div hx-ext="client-side-templates">
    <div mustache-template="foo" hx-sse="connect:/ticker swap:symbol" class="container"> 
   
    </div>
    <template id="foo">
      {{#.}}
         {{symbol}} - {{price}}
      {{/.}}
    </template>
</div>
```

Please note, the example above uses mustache templates, but client-side-templates extension also supports handlbars and nunjucks, so feel free to pick. Check out the docs for more info and a full page example  https://htmx.org/extensions/client-side-templates/

**Reply by karapy** [reply]
> 

:(
Because I am using FastAPI for backend and this backend is using jinja2 as a template engine, it seems they are conflicting with each other that's why i got into problem, is there any way?

**Reply by Renerick** [reply]
https://jinja.palletsprojects.com/en/3.0.x/templates/#escaping 

Refer to the link above. Seem like you can use `{{ '{{something}}' }}` technique to escape small strings and `raw` blocks for larger fragments

**Reply by Renerick** [reply]
I sent you docs for jinja 3, but jinja 2 should work exactly the same https://jinja.palletsprojects.com/en/2.11.x/templates/#escaping

**Reply by karapy** [reply]
here is how i used template engine

![image](https://user-images.githubusercontent.com/102681330/176769532-132dda1b-ef5e-42bf-bfe4-6ba74ad70a3a.png)

here is the result:

![image](https://user-images.githubusercontent.com/102681330/176769273-08240230-699f-4114-a1ff-bc04d5247b0b.png)

and here is the SSE's link itself:
![image](https://user-images.githubusercontent.com/102681330/176770410-eeefc763-fec8-4c09-8238-800df7ec06b2.png)


What did I wrong?

**Reply by Renerick** [reply]
You've put your template inside of your target element. After the first swap, the template is overwritten and client-side-templates extension is unable to use it. Put it outside, like in the example I provided

```html
<div mustache-template="foo" hx-sse="connect:/ticker swap:message" class="container"> 

</div> <!-- target ends here -->

<template id="foo"> <!-- template is outside -->
  {{#.}}
     {{symbol}} - {{price}}
  {{/.}}
</template>
```

**Reply by karapy** [reply]
I did the same Fastapi uses jinja2, I used 
`https://jinja.palletsprojects.com/en/2.11.x/templates/#escaping`
the be able to use mustache in htmx but still the same issue :(

```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  
  <script src="https://unpkg.com/htmx.org"></script>
  <script src="https://unpkg.com/htmx.org/dist/ext/client-side-templates.js"></script>
  <script src="https://unpkg.com/mustache@latest"></script>

  <title>JS Bin</title>
</head>
<body>
    {% raw %}
    
    <div mustache-template="foo" hx-sse="connect:/ticker swap:message" class="container"  hx-target="#output"> 

    </div>
    
    
        <!-- target ends here -->
    
    
    <template id="foo"> <!-- template is outside -->
              
          {{#.}}
          {{symbol}} - {{price}}
          {{/.}}
        
    </template>  
   

        <div id="output">
            
        </div>
        {% endraw %}
    
</body>
</html>
```

**Reply by karapy** [reply]
Sovled finally!!! and even didnt need escape anything from jinja2
```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>JS Bin</title>
  <script src="https://unpkg.com/htmx.org"></script>
  <script src="https://unpkg.com/htmx.org/dist/ext/client-side-templates.js"></script>
  <script src="https://unpkg.com/mustache@latest"></script>
</head>
<body>
  <div hx-ext="client-side-templates">
    <div mustache-template="foo" hx-sse="connect:/ticker swap:symbol" class="container"> 
    he here?
    </div>

    <template id="foo">
        {{#.}}
           <li>{{symbol}} - {{price}} </li>
        {{/.}}
    </template>

  </div>
</body>
</html>
```
